### DISCUSSION TITLE
Problem about generating group constants using mlotres_xs.py

### DISCUSSION THREAD

**Discussion by js-jixu** [arfc/moltres#204] [Q&A Getting started] [answered]
Hi, everyone. I want to use OpenMC to generate group constants for Moltres using `moltres_xs.py` , the number of groups is `7` and the number of precursor groups is `6` .

Following what Godiva said in the README.md, I wrote three OpenMC input files (with different temperatures, jixu_openmc_900 and so on) including the ` openmc_xs.generate_openmc_tallies_xml() ` function. But I don't know how to set the parameters of `domain_dict` part . The example only uses one material but I want to use three materials, which are fuel(id=1) ,  clad(id=2) and cool(id=3). I don't know how to set other materials in domain_dict part(900k.py, line 85-92). Should I write code like this?
```
tallies_file = openmc.Tallies()
domain_dict = openmc_xs.generate_openmc_tallies_xml(
    [1e-5,2.26,3.043e2,2.035e3,9.12e3,6.74e4,1.36e6,2e7],
    list(range(1, 9)),
    [fuel] [clad] [cool],
    [fuel.id] [clad.id] [cool.id],
    tallies_file,
)
```
Should I replace `[fuel.id]` with `1` ?

Besides that, when I run the script, there is an DiffusionCoefficient error: 
![DiffusionCoefficientError](https://user-images.githubusercontent.com/99813361/169804247-98124530-e9c8-4ab6-beb5-c78b6a5afd21.png)
Am I missing any necessary parameters?

If I get OpenMC output files: `statepoint_900_openmc.100.h5` , `summary_900.h5` and so on,  should I write moltres_xs.py input file `jixu_openmc.inp` like this?
```
[TITLE]
  jixu_openmc.json

[MAT]
  3
  fuel clad cool

[BRANCH]
  3
  fuel clad cool            900 1 1 1 2 3 1
  fuel clad cool            1000 1 1 1 2 3 1
  fuel clad cool            1100 1 1 1 2 3 1

[FILES]
  3
  statepoint_900_openmc.100.h5 openmc jixu_openmc_900.py summary_900.h5
  statepoint_1000_openmc.100.h5 openmc jixu_openmc_1000.py summary_1000.h5
  statepoint_1100_openmc.100.h5 openmc jixu_openmc_1100.py summary_1100.h5
```
The `1 2 3` in `fuel clad cool            900 1 1 1 2 3 1` represents the ids of fuel, clad and cool.

The OpenMC input files for different tempratures, moltres_xs.py, README.txt and jixu_openmc.inp for moltres_xs.py are attached. Any advice or correction are appreciated.
[Moltres_Discussion.zip](https://github.com/arfc/moltres/files/8754502/Moltres_Discussion.zip)

**Comment by gwenchee** [comment, marked as answer]
The DiffusionCoefficient function in OpenMC is only available in v0.12.1 and newer. 

As for the domain_dict, if you have more than one material, your input would look like: 
```
domain_dict = openmc_xs.generate_openmc_tallies_xml(
    [1e-5, 748.5, 5.5308e3, 24.7875e3, 0.4979e6, 2.2313e6, 20e6],
    list(range(1, 9)),
    [fuel_mat, fuel_mat2],
    [fuel_mat.id, fuel_mat2.id],
    tallies_file,
)
```

> Should I replace [fuel.id] with 1 ? 

No. It should be the fuel's id. See OpenMC's Material properties: https://docs.openmc.org/en/latest/pythonapi/generated/openmc.Material.html 

> If I get OpenMC output files: statepoint_900_openmc.100.h5 , summary_900.h5 and so on, should I write moltres_xs.py input file jixu_openmc.inp like this?

You should not repeat 1100K. It looks good otherwise

**Reply by js-jixu** [reply]
Thank you a lot for the helpful reply, Miss Chee. I spent a long time on the problem of DiffusionCofficient.

I don't understand the meaning of `You should not repeat 1100K` .  When I want to interpolate data with three temperatures, shouldn't I list the three temperatures (900K, 1000K, 1100K) separately?  Besides that, I would like to know if the [BRANCH] section in my INP file is written correctly.

Thanks again!

**Reply by gwenchee** [reply]
Oh, my bad. I read both 1000K and 1100K as 1100K. It looks good as is then. 

In the documentation, this is the format of the [BRANCH]: 
```
[BRANCH]
  *number of branches*
  *material name* *temperature* *file index* *burnup index* *universe index* *branch index*
```

Since you have 3 materials in each file, you should have 9 branches. Each material in each file should be its own line in branch. An example: 

```
[BRANCH]
  4
  fuel_mat            900 1 1 1 1
  fuel_mat2          900 1 1 2 1
  fuel_mat            1200 2 1 1 1
  fuel_mat2          1200 2 1 2 1
```
The universe index refers to the material's id. Thus, in this example, I'm assuming the fuel_mat.id = 1 and fuel_mat2.id = 2.

**Reply by js-jixu** [reply]
Thank you, Miss Chee. I finally understand the usage of moltres_xs.py !

**Reply by js-jixu** [reply]
Hi, Miss Chee. I've encountered another problem when I want to generate JSON file.

I installed OpenMC yesterday according to [the source code installation tutorial](https://docs.openmc.org/en/stable/quickinstall.html) on the official website. The version of OpenMC is `0.13.1-dev` and the version of python is `3.8` . When I tried to use the following command to generate JSON file by `godiva_openmc.inp` 
```
python3 moltres_xs.py godiva_openmc.inp
```

 It reported an error: 

![error](https://user-images.githubusercontent.com/99813361/170810349-a00add09-9930-4d78-a9c0-3729b7ffcbb4.png)

The version of pandas is `1.4.2` . But I don't think the root of error is the version of pandas, because the error report is same when I use other version of pandas.

But if I substitute `legendre` with `mean` (moltres_xs.py, line128), it runs sucessfully while the JSON file has no `DFFCOEF` , just like the following picture shows:

![mean](https://user-images.githubusercontent.com/99813361/170817851-cfebc50f-c178-4b2f-a9f6-6a5744694ce0.png)

I want to know how to solve it, should I use another version of OpenMC (preferably not, it took me two days to install it due to network issues)? All files mentioned in the error report are attached. Besides that, files used to generate JSON file are also attached in `godiva` .

[files.zip](https://github.com/arfc/moltres/files/8790501/files.zip)

**Reply by smpark7** [reply]
It seems that OpenMC changed how they handle and display the [diffusion coefficients](https://github.com/openmc-dev/openmc/pull/1909). Unfortunately, we're currently swamped with other projects and so we won't be able to work on the fix. In the meantime, you can either install an earlier OpenMC version before they made this change, or edit the moltres_xs.py file specifically for your current OpenMC version.

**Reply by gwenchee** [reply]
I suggest installing openmc v-0.12.2. I'm using it now and it works with DiffusionCoefficient. @js-jixu

**Reply by js-jixu** [reply]
I made a little modification on the original code and it is now compatible with the latest version of openmc(v-0.13.1). But it still has a bug, that is, it cannot generate group constants for two or more temperatures at the same time. Anyway, it works for me now.

![modification](https://user-images.githubusercontent.com/99813361/171471247-73978837-5046-4027-a73a-539f02ffcf19.png)
[moltres_xs.zip](https://github.com/arfc/moltres/files/8817803/moltres_xs.zip)

**Reply by gwenchee** [reply]
Excellent. You're welcome to make a PR into moltres_xs.py if you fix the bug :)

**Reply by smpark7** [reply]
Note that any change should ensure backward compatibility to the older 0.12.x versions by checking which OpenMC version the user ran to generate the statefiles.

**Reply by js-jixu** [reply]
Yeah, to be on the safe side, I still use openmc `v.0.12.2` .
